lua/autorun/client/cl_devtools.lua

```diff
@@ -1,144 +1,165 @@
-        else
-            print("Include handled by file")
+        -- remove the variaables from the global.
+        for key,_ in pairs(environmentVars) do
+            environment[key] = nil
         end
     end
 
-    -- only replace includes that don't use variables
-    RunString(string.gsub(code, "include%s*%(%s*[\'\"].-[\'\"]%s*%)", ""), "swampDevTools")
-
-    return filename
-end
-
-local function recurseRefresh(root, times, curPath)
-    local files, folders = file.Find(root .. curPath .. "*", "MOD")
+    local fileTime = file.Time
 
-    for k, v in ipairs(files) do
-        if (not string.EndsWith(v, ".lua") or string.StartWith(v, "sv_") or (v == "init.lua")) then continue end
-        local newtime = file.Time(root .. curPath .. v, "MOD")
+    function refreshFolder(subDir)
+        local fileNames, folders = file.Find(subDir .. "/*", "MOD")
+        local fileTimes = SWAMP_DEV.fileTimes
+        for _,fileName in ipairs(fileNames) do
+            local filePath = subDir .. "/" .. fileName
+            if not fileName:EndsWith(".lua") then continue end
 
-        -- check if file has updated
-        if times[k] and (newtime ~= times[k]) then
-            print("File refreshed: " .. root .. curPath .. refreshFile(root .. curPath .. v, v))
-        end
+            local fTime = fileTime(filePath, "MOD")
 
-        times[k] = newtime
-    end
+            if not fileTimes[filePath] then
+                fileTimes[filePath] = fTime
+                continue
+            else
+                if fileTimes[filePath] == fTime then continue end
+            end
 
-    for _, v in ipairs(folders) do
-        if (v == "server") then continue end
+            local wasFoundDuringLookup = false
+            for structName, func in pairs(SWAMP_DEV.structEnvironments) do
+                local structLookupPath = "lua/" .. structName
 
-        if (not times[v]) then
-            times[v] = {}
-        end
+                if filePath:find(structLookupPath) then
+                    local className = fileName:sub(1,-5) -- remove the .lua extension.
+                    loadFile(filePath, func(className) or func(subDir:gsub(".*/", "")))
+                    wasFoundDuringLookup = true
+                end
+            end
 
-        -- defining struct
-        if SWAMP_DEV.structTypes[v] then
-            SWAMP_DEV.itemStruct = v
+            if not wasFoundDuringLookup then 
+                loadFile(filePath)
+            end
+            fileTimes[filePath] = fTime
         end
 
-        times[v] = recurseRefresh(root, times[v], curPath .. v .. "/")
-
-        if (curPath == "") then
-            SWAMP_DEV.itemStruct = nil
+        for _,folder in ipairs(folders) do
+            refreshFolder(subDir .. "/" .. folder)
         end
     end
-
-    return times
 end
 
 -- toggle dev editing. Root is addons/contrib
 concommand.Add("dev", function()
     if (LocalPlayer():GetRank() <= 0) then return end
```

Comment:
Confirm that this is actually not going to open up dev tools to donators or something.

Reply:
Rank is only for staff. However I was actually thinking we should have a hardcoded steamID list in here so devs don't have to be staff to use these tools.

Reply:
Maybe it'd be better to put this table in a different location and made global so we can reference it in other systems. Perhaps if you wanted a contributor tag on player overheads or something, etc.